packages/core/contracts/oracle/implementation/SafetyModule.sol

```diff
@@ -0,0 +1,119 @@
+// SPDX-License-Identifier: AGPL-3.0-only
+pragma solidity 0.8.16;
+
+import "./Finder.sol";
+import "./GovernorV2.sol"; //todo: swap this to using an interface.
+import "./Constants.sol";
+import "../interfaces/StakerInterface.sol";
+import "./AdminIdentifierLib.sol";
+import "../../common/implementation/Lockable.sol";
+import "@openzeppelin/contracts/access/Ownable.sol";
+import "@openzeppelin/contracts/token/ERC20/IERC20.sol";
+import "@openzeppelin/contracts/token/ERC20/utils/SafeERC20.sol";
+
+contract SafteyModule is Ownable, Lockable {
+    using SafeERC20 for IERC20;
+    IERC20 public immutable token;
+    uint256 public bond;
+    GovernorV2 public immutable governor;
+    Finder public immutable finder;
+
+    struct EmergencyProposal {
+        address proposer;
+        uint64 time;
+        uint256 lockedBond;
+        uint256 cumulativeSignaled;
+        mapping(address => uint256) accountSignaled;
+        bool ratified;
+    }
+    EmergencyProposal[] public emergencyProposals;
+
+    uint256 emergencyActionThreshold = 0.5e18;
+
+    address public emergencyActionAuthenticator;
+
+    event BondSet(uint256 bond);
+    event ProposalResolved(uint256 indexed id, bool success);
+
+    constructor(
+        IERC20 _token, //todo: swap this to having the "voting token" in the finder and remove this.
+        uint256 _bond,
+        GovernorV2 _governor, // todo: add the governor to the finder and remove this.
+        Finder _finder
+    ) {
+        token = _token;
+        governor = _governor;
+        finder = _finder;
+        setBond(_bond);
+        transferOwnership(address(_governor));
+    }
+
+    function proposeEmergencyAction(GovernorV2.Transaction[] memory transactions)
+        external
+        nonReentrant()
+        returns (uint256)
+    {
+        uint256 id = emergencyProposals.length - 1;
+        emergencyProposals.push();
+        token.safeTransferFrom(msg.sender, address(this), bond);
+        emergencyProposals[id].proposer = msg.sender;
+        emergencyProposals[id].lockedBond = bond;
+        emergencyProposals[id].time = uint64(getCurrentTime());
+
+        governor.proposeEmergencyAction(transactions);
+        return id;
+    }
+
+    function signalOnEmergencyProposal(uint256 id) public {
+        EmergencyProposal storage emergencyProposal = emergencyProposals[id];
+
+        require(emergencyProposal.accountSignaled[msg.sender] == 0);
+
+        emergencyProposal.accountSignaled[msg.sender] = getVoterStake(msg.sender);
+        emergencyProposal.cumulativeSignaled += emergencyProposal.accountSignaled[msg.sender];
+    }
```

Comment:
If users are able to unstake, they could replicate their votes, right? I could stake with account1, signal, withdraw, restake with a second account, signal, etc.

Reply:
ye, you right. in the implementation where this was within the voting, contract we'd block a voter from unstaking if they'd signed on an emergency action. options would be to block the staker from unstaking within the voting contract if they'd signed on an emergency action?

Reply:
I've updated the proposal with this minimal change. basically, we now track a counter of how many proposals an account is signaled on. you can only Unstake if this is set to 0, thereby forcing you to cancel your signal on everything before you can unstake. This does add a small gas cost to executing an unstake but it should not be too much as there are only two small read intra-contract calls (call to the finder and a call to the safety module).